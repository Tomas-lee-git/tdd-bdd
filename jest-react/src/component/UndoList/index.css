.undo-list {
  display: flex;
  justify-content: center;
  align-items: center;
}
.undo-list-container {
  width: 500px;
}
.undo-list-title {
  font-size: 20px;
  font-weight: bold;
  display: flex;
  justify-content: space-between;
}
.undo-list-counter {
  font-size: 14px;
  color: #666;
  background-color: #e6e6fa;
  border-radius: 50%;
  line-height: 20px;
  height: 20px;
  padding: 0 5px;
  text-align: center;
}
.undo-list-content {
  list-style-type: none;
}
.undo-list-item {
  line-height: 24px;
  height: 24px;
  font-size: 24px;
  margin-bottom: 10px;
  background-color: #fff;
  border-left: 3px solid #629a9c;
  text-indent: 10px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  border-radius: 3px;
}
.delete-item {
  margin-right: 10px;
  background: #e6e6ea;
  text-align: center;
  width: 20px;
  height: 20px;
  line-height: 20px;
  border-radius: 10px;
  text-indent: 0px;
}
