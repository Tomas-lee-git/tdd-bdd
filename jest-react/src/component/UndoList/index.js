import React, { Component } from 'react';
import './index.css';
class UndoList extends Component {
  handleInputItemChange = (e, index) => {
    const { value } = e.target;
    const { handleModifiedItem } = this.props;
    handleModifiedItem(index, value);
  };

  handleClickWithEnter = e => {
    const { keyCode } = e;
    const { handleBlur } = this.props;
    if (keyCode === 13) {
      handleBlur();
    }
  };

  render() {
    const { undoList, deleteItem, handleFocus, handleBlur } = this.props;
    return (
      <div className="undo-list">
        <div className="undo-list-container">
          <div className="undo-list-title">
            正在进行
            <span className="undo-list-counter" data-test="counter">
              {undoList.length}
            </span>
          </div>
          <ul className="undo-list-content">
            {undoList.map((item, index) => (
              <li
                key={`${index}-${item}`}
                className="undo-list-item"
                data-test="undoItem"
                onClick={() => {
                  handleFocus(index);
                }}
                onBlur={() => {
                  handleBlur();
                }}
              >
                {item.isFocus ? (
                  <input
                    autoFocus
                    data-test="inputItem"
                    value={item.value}
                    className="undo-list-input-item"
                    onChange={e => this.handleInputItemChange(e, index)}
                    onKeyUp={e => {
                      this.handleClickWithEnter(e, index);
                    }}
                  ></input>
                ) : (
                  <div data-test="divItem" className="undo-list-div-item">
                    {item.value}
                  </div>
                )}
                <div
                  className="delete-item"
                  data-test="deleteItem"
                  onClick={e => {
                    e && e.stopPropagation();
                    deleteItem(index);
                  }}
                >
                  -
                </div>
              </li>
            ))}
          </ul>
        </div>
      </div>
    );
  }
}
export default UndoList;
