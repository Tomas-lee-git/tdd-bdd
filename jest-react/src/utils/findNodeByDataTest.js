const findNodeByDataTest = (wrapper, tag) => wrapper.find(`[data-test='${tag}']`);
export { findNodeByDataTest };
