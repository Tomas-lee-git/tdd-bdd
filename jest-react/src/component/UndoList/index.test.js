import React from 'react';
import { shallow } from 'enzyme';

import UndoList from './index';
import { findNodeByDataTest } from '../../utils/findNodeByDataTest';

describe('UndoList component test', () => {
  describe('dom test', () => {
    test('counter text should be 0, and undoList length should be 0', () => {
      const wrapper = shallow(<UndoList undoList={[]} />);
      const counter = findNodeByDataTest(wrapper, 'counter');
      expect(counter.text()).toEqual('0');
      const undoList = findNodeByDataTest(wrapper, 'undoList');
      expect(undoList.length).toEqual(0);
    });

    test('counter text should be 3, undo list length should be 3', () => {
      const mockUndoList = [
        { isFocus: false, value: 'hello' },
        { isFocus: false, value: 'hello' },
        { isFocus: false, value: 'hello' },
      ];
      const wrapper = shallow(<UndoList undoList={mockUndoList} />);
      const counter = findNodeByDataTest(wrapper, 'counter');
      expect(counter.text()).toEqual('3');
      const undoList = findNodeByDataTest(wrapper, 'undoList');
      expect(undoList.length).toEqual(3);
    });
    test("should have one inputItem when item's isFocus is true", () => {
      const mockUndoList = [
        { isFocus: false, value: 'hello' },
        { isFocus: true, value: 'hello' },
        { isFocus: false, value: 'hello' },
      ];
      const wrapper = shallow(<UndoList undoList={mockUndoList} />);
      const inputItems = findNodeByDataTest(wrapper, 'inputItem');
      expect(inputItems.length).toEqual(1);
    });
  });

  describe('delete function test', () => {
    test('should have corresponding delete button when list exist', () => {
      const mockUndoList = [
        { isFocus: false, value: 'hello' },
        { isFocus: false, value: 'hello' },
        { isFocus: false, value: 'hello' },
      ];
      const wrapper = shallow(<UndoList undoList={mockUndoList} />);
      const deleteItems = findNodeByDataTest(wrapper, 'deleteItem');
      expect(deleteItems.length).toEqual(3);
    });

    test('delete function should be called after click delete btn', () => {
      const mockUndoList = [
        { isFocus: false, value: 'hello' },
        { isFocus: false, value: 'hello' },
        { isFocus: false, value: 'hello' },
      ];
      const mockDeleteItemFn = jest.fn();
      const wrapper = shallow(
        <UndoList undoList={mockUndoList} deleteItem={mockDeleteItemFn} />
      );
      const deleteItems = findNodeByDataTest(wrapper, 'deleteItem');
      const index = 1;
      deleteItems.at(index).simulate('click');
      expect(mockDeleteItemFn).toHaveBeenCalledWith(index);
      expect(mockDeleteItemFn).toHaveBeenCalledTimes(1);
    });
  });

  describe('click listItem test', () => {
    test('handleFocus fn should be called after click listItem', () => {
      const mockUndoList = [
        { isFocus: false, value: 'hello' },
        { isFocus: false, value: 'hello' },
        { isFocus: false, value: 'hello' },
      ];
      const mockHandleFocus = jest.fn();
      const wrapper = shallow(
        <UndoList undoList={mockUndoList} handleFocus={mockHandleFocus} />
      );
      const listItems = findNodeByDataTest(wrapper, 'undoList');
      const index = 1;
      listItems.at(index).simulate('click');
      expect(mockHandleFocus).toHaveBeenCalledWith(index);
      expect(mockHandleFocus).toHaveBeenCalledTimes(1);
    });
  });
});
