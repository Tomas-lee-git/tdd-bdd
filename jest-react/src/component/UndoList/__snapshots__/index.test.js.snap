// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UndoList component test dom test snapshot test 1`] = `
<div
  className="undo-list"
>
  <div
    className="undo-list-container"
  >
    <div
      className="undo-list-title"
    >
      正在进行
      <span
        className="undo-list-counter"
        data-test="counter"
      >
        0
      </span>
    </div>
    <ul
      className="undo-list-content"
    />
  </div>
</div>
`;
